### ISSUE TITLE
Soft-yanking

### ISSUE THREAD

**Issue by pirj** [rubygems/rfcs#34]
### Problem

Occasionally, when maintainers publish new gem versions, they [make mistakes](https://en.wikipedia.org/wiki/List_of_Latin_phrases_(E)#errare_humanum_est).

Examples:

1. `rails` Ruby version constraint mistake that broke 5.2.4.3-5.2.4.5 on Ruby 2.2. See https://github.com/rails/rails/blob/v5.2.4.3/activesupport/lib/active_support/cache/redis_cache_store.rb#L323
It [has been fixed in May 2020](https://github.com/rails/rails/commit/a8844200031e40b9a452bc9374de1f7b9440ac7f), but only released nearly a year later in [Rails 5.2.4.6 (May 2021)](https://github.com/rails/rails/blob/v5.2.4.6/activesupport/lib/active_support/cache/redis_cache_store.rb#L323).

2. `rspec-rails` Ruby version constraint mistake that [broke `rspec-rails` on Ruby 2.2](https://github.com/rspec/rspec-rails/pull/2319).

3. `diff-lcs` issue [with older Ruby versions](https://github.com/halostatue/diff-lcs/issues/63)

4. `cucumber` 4.0.0 broke compatibility due to `diff-lcs` dependency, [fixed in 4.0.1](https://github.com/cucumber/cucumber-ruby/pull/1430) by pinning `diff-lcs` version to `~> 1.3`.

### Suggestion: Soft-yank

What soft-yanking means?

#### Gem maintainer scenario

The maintainer can soft-yank a gem version, just like they can yank it:
```sh
gem soft-yank GEM -v VERSION [-p PLATFORM] [--key KEY_NAME] [--host HOST]
```

#### Server/CI scenario

It remains possible to install the soft-yanked version of a gem with `bundle install` from Gemfile.lock.
Bundler emits a warning.

#### Developer scenario

Bundler excludes soft-yanked versions from dependency resolution.
`bundle update`/`bundle lock` show an error, just like for a yanked gem version or a removed gem.

### Could things have gone better?

`rspec-rails` 4.0.0 could have been soft-yanked.
`cucumber` 4.0.0 could have been soft-yanked.
`diff-lcs` 1.4.3 could have been soft-yanked.

I have no such certainty regarding Rails, since it took a year to release the fix.

### Misc

Related: https://github.com/rubygems/rubygems/issues/1506#issuecomment-188472423

#26 is semi-related, a proposal to prevent the only cause I'm practically aware of, weak Ruby version constraint. There might be others, like adding extra runtime dependencies, but I have not seen this in the wild.

cc @halostatue @JonRowe @marcandre @mattwynne @aslakhellesoy.

**Comment by marcandre** [comment]
> If all you need to do to fix this problem currently is to lock rspec-rails to ~> 3.9 it is still super simple solution (considering how old Ruby is used here).

Right, and the proposal here does not solve anything for `rspec-rails` 4.0 or Ruby 2.2, of course, and that's fine :-) It only aims to make it possible and easy in the future to remove invalid gem manifests from the circulation, without too much impact.

**Comment by marcandre** [comment]
BTW, I was curious about the equivalent in `elixir` world. `hex` appears to have no `yank` equivalent, but can ["retire"](https://hexdocs.pm/hex/Mix.Tasks.Hex.Retire.html#content) a package; that's informational only and will still resolve.

**Comment by pirj** [comment]
Any conclusion? I can send an RFC PR, but would like to know that it's preliminarily agreed on.
I see two options:

1. Introduce soft-yanking as a new feature.
2. Introduce soft-yanking for a wide audience, and restrict hard-yanking to the RubyGems team.

**Comment by jchestershopify** [comment]
As an alternative, can a hard yank instead include a suggested alternate version?

**Comment by pirj** [comment]
[Another case](https://github.com/rubocop/rubocop-rspec/commit/eed7439be0c90c7d79e264240f91ad0a924d5a80#r96342863) where soft-yanking or yanking with an alternate version would help mitigate a loose dependency mistake.
